// @base-ui-components/react/esm/progress/label/ProgressLabel.js
'use client';

import * as React from 'react';
import PropTypes from 'prop-types';
import { useBaseUiId } from '../../utils/useBaseUiId.js';
import { useRenderElement } from '../../utils/useRenderElement.js';
import { useEnhancedEffect } from '../../utils/useEnhancedEffect.js';
import { useProgressRootContext } from '../root/ProgressRootContext.js';
import { progressStyleHookMapping } from '../root/styleHooks.js';
/**
 * An accessible label for the progress bar.
 * Renders a `<span>` element.
 *
 * Documentation: [Base UI Progress](https://base-ui.com/react/components/progress)
 */
const ProgressLabel = /*#__PURE__*/React.forwardRef(function ProgressLabel(componentProps, forwardedRef) {
  const {
    render,
    className,
    id: idProp,
    ...elementProps
  } = componentProps;
  const id = useBaseUiId(idProp);
  const {
    setLabelId,
    state
  } = useProgressRootContext();
  useEnhancedEffect(() => {
    setLabelId(id);
    return () => setLabelId(undefined);
  }, [id, setLabelId]);
  const renderElement = useRenderElement('span', componentProps, {
    state,
    ref: forwardedRef,
    props: [{
      id
    }, elementProps],
    customStyleHookMapping: progressStyleHookMapping
  });
  return renderElement();
});
export { ProgressLabel };
process.env.NODE_ENV !== "production" ? ProgressLabel.propTypes /* remove-proptypes */ = {
  // ┌────────────────────────────── Warning ──────────────────────────────┐
  // │ These PropTypes are generated from the TypeScript type definitions. │
  // │ To update them, edit the TypeScript types and run `pnpm proptypes`. │
  // └─────────────────────────────────────────────────────────────────────┘
  /**
   * @ignore
   */
  children: PropTypes.node,
  /**
   * CSS class applied to the element, or a function that
   * returns a class based on the component’s state.
   */
  className: PropTypes.oneOfType([PropTypes.func, PropTypes.string]),
  /**
   * @ignore
   */
  id: PropTypes.string,
  /**
   * Allows you to replace the component’s HTML element
   * with a different tag, or compose it with another component.
   *
   * Accepts a `ReactElement` or a function that returns the element to render.
   */
  render: PropTypes.oneOfType([PropTypes.element, PropTypes.func])
} : void 0;
