// @base-ui-components/react/cjs/select/arrow/SelectArrowDataAttributes.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SelectArrowDataAttributes = void 0;
var _popupStateMapping = require("../../utils/popupStateMapping");
let SelectArrowDataAttributes = exports.SelectArrowDataAttributes = function (SelectArrowDataAttributes) {
  /**
   * Present when the select popup is open.
   */
  SelectArrowDataAttributes[SelectArrowDataAttributes["open"] = _popupStateMapping.CommonPopupDataAttributes.open] = "open";
  /**
   * Present when the select popup is closed.
   */
  SelectArrowDataAttributes[SelectArrowDataAttributes["closed"] = _popupStateMapping.CommonPopupDataAttributes.closed] = "closed";
  /**
   * Present when the anchor is hidden.
   */
  SelectArrowDataAttributes[SelectArrowDataAttributes["anchorHidden"] = _popupStateMapping.CommonPopupDataAttributes.anchorHidden] = "anchorHidden";
  /**
   * Indicates which side the select is positioned relative to the trigger.
   * @type {'none' | 'top' | 'bottom' | 'left' | 'right' | 'inline-end' | 'inline-start'}
   */
  SelectArrowDataAttributes["side"] = "data-side";
  /**
   * Present when the select arrow is uncentered.
   */
  SelectArrowDataAttributes["uncetered"] = "data-uncentered";
  return SelectArrowDataAttributes;
}({});
