// @base-ui-components/react/esm/popover/arrow/PopoverArrowDataAttributes.js
import { CommonPopupDataAttributes } from '../../utils/popupStateMapping.js';
export let PopoverArrowDataAttributes = function (PopoverArrowDataAttributes) {
  /**
   * Present when the popup is open.
   */
  PopoverArrowDataAttributes[PopoverArrowDataAttributes["open"] = CommonPopupDataAttributes.open] = "open";
  /**
   * Present when the popup is closed.
   */
  PopoverArrowDataAttributes[PopoverArrowDataAttributes["closed"] = CommonPopupDataAttributes.closed] = "closed";
  /**
   * Present when the anchor is hidden.
   */
  PopoverArrowDataAttributes[PopoverArrowDataAttributes["anchorHidden"] = CommonPopupDataAttributes.anchorHidden] = "anchorHidden";
  /**
   * Indicates which side the popup is positioned relative to the trigger.
   * @type {'top' | 'bottom' | 'left' | 'right' | 'inline-end' | 'inline-start'}
   */
  PopoverArrowDataAttributes["side"] = "data-side";
  /**
   * Present when the popover arrow is uncentered.
   */
  PopoverArrowDataAttributes["uncetered"] = "data-uncentered";
  return PopoverArrowDataAttributes;
}({});
