// @base-ui-components/react/esm/preview-card/arrow/PreviewCardArrowDataAttributes.js
import { CommonPopupDataAttributes } from '../../utils/popupStateMapping.js';
export let PreviewCardArrowDataAttributes = function (PreviewCardArrowDataAttributes) {
  /**
   * Present when the preview card is open.
   */
  PreviewCardArrowDataAttributes[PreviewCardArrowDataAttributes["open"] = CommonPopupDataAttributes.open] = "open";
  /**
   * Present when the preview card is closed.
   */
  PreviewCardArrowDataAttributes[PreviewCardArrowDataAttributes["closed"] = CommonPopupDataAttributes.closed] = "closed";
  /**
   * Present when the anchor is hidden.
   */
  PreviewCardArrowDataAttributes[PreviewCardArrowDataAttributes["anchorHidden"] = CommonPopupDataAttributes.anchorHidden] = "anchorHidden";
  /**
   * Indicates which side the preview card is positioned relative to the trigger.
   * @type {'top' | 'bottom' | 'left' | 'right' | 'inline-end' | 'inline-start'}
   */
  PreviewCardArrowDataAttributes["side"] = "data-side";
  /**
   * Present when the preview card arrow is uncentered.
   */
  PreviewCardArrowDataAttributes["uncetered"] = "data-uncentered";
  return PreviewCardArrowDataAttributes;
}({});
