// @base-ui-components/react/esm/meter/indicator/MeterIndicator.js
'use client';

import * as React from 'react';
import PropTypes from 'prop-types';
import { mergeProps } from '../../merge-props/index.js';
import { useComponentRenderer } from '../../utils/useComponentRenderer.js';
import { useMeterRootContext } from '../root/MeterRootContext.js';
const EMPTY = {};
/**
 * Visualizes the position of the value along the range.
 * Renders a `<div>` element.
 *
 * Documentation: [Base UI Meter](https://base-ui.com/react/components/meter)
 */
const MeterIndicator = /*#__PURE__*/React.forwardRef(function MeterIndicator(props, forwardedRef) {
  const {
    render,
    className,
    ...otherProps
  } = props;
  const {
    percentageValue
  } = useMeterRootContext();
  const getStyles = React.useCallback(() => {
    return {
      insetInlineStart: 0,
      height: 'inherit',
      width: `${percentageValue}%`
    };
  }, [percentageValue]);
  const propGetter = React.useCallback((externalProps = {}) => mergeProps({
    style: getStyles()
  }, externalProps), [getStyles]);
  const {
    renderElement
  } = useComponentRenderer({
    propGetter,
    render: render ?? 'div',
    state: EMPTY,
    className,
    ref: forwardedRef,
    extraProps: otherProps
  });
  return renderElement();
});
export { MeterIndicator };
process.env.NODE_ENV !== "production" ? MeterIndicator.propTypes /* remove-proptypes */ = {
  // ┌────────────────────────────── Warning ──────────────────────────────┐
  // │ These PropTypes are generated from the TypeScript type definitions. │
  // │ To update them, edit the TypeScript types and run `pnpm proptypes`. │
  // └─────────────────────────────────────────────────────────────────────┘
  /**
   * @ignore
   */
  children: PropTypes.node,
  /**
   * CSS class applied to the element, or a function that
   * returns a class based on the component’s state.
   */
  className: PropTypes.oneOfType([PropTypes.func, PropTypes.string]),
  /**
   * Allows you to replace the component’s HTML element
   * with a different tag, or compose it with another component.
   *
   * Accepts a `ReactElement` or a function that returns the element to render.
   */
  render: PropTypes.oneOfType([PropTypes.element, PropTypes.func])
} : void 0;
