// @base-ui-components/react/esm/select/arrow/SelectArrowDataAttributes.js
import { CommonPopupDataAttributes } from '../../utils/popupStateMapping.js';
export let SelectArrowDataAttributes = function (SelectArrowDataAttributes) {
  /**
   * Present when the select popup is open.
   */
  SelectArrowDataAttributes[SelectArrowDataAttributes["open"] = CommonPopupDataAttributes.open] = "open";
  /**
   * Present when the select popup is closed.
   */
  SelectArrowDataAttributes[SelectArrowDataAttributes["closed"] = CommonPopupDataAttributes.closed] = "closed";
  /**
   * Present when the anchor is hidden.
   */
  SelectArrowDataAttributes[SelectArrowDataAttributes["anchorHidden"] = CommonPopupDataAttributes.anchorHidden] = "anchorHidden";
  /**
   * Indicates which side the select is positioned relative to the trigger.
   * @type {'none' | 'top' | 'bottom' | 'left' | 'right' | 'inline-end' | 'inline-start'}
   */
  SelectArrowDataAttributes["side"] = "data-side";
  /**
   * Present when the select arrow is uncentered.
   */
  SelectArrowDataAttributes["uncetered"] = "data-uncentered";
  return SelectArrowDataAttributes;
}({});
