// @base-ui-components/react/cjs/preview-card/arrow/PreviewCardArrowDataAttributes.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.PreviewCardArrowDataAttributes = void 0;
var _popupStateMapping = require("../../utils/popupStateMapping");
let PreviewCardArrowDataAttributes = exports.PreviewCardArrowDataAttributes = function (PreviewCardArrowDataAttributes) {
  /**
   * Present when the preview card is open.
   */
  PreviewCardArrowDataAttributes[PreviewCardArrowDataAttributes["open"] = _popupStateMapping.CommonPopupDataAttributes.open] = "open";
  /**
   * Present when the preview card is closed.
   */
  PreviewCardArrowDataAttributes[PreviewCardArrowDataAttributes["closed"] = _popupStateMapping.CommonPopupDataAttributes.closed] = "closed";
  /**
   * Present when the anchor is hidden.
   */
  PreviewCardArrowDataAttributes[PreviewCardArrowDataAttributes["anchorHidden"] = _popupStateMapping.CommonPopupDataAttributes.anchorHidden] = "anchorHidden";
  /**
   * Indicates which side the preview card is positioned relative to the trigger.
   * @type {'top' | 'bottom' | 'left' | 'right' | 'inline-end' | 'inline-start'}
   */
  PreviewCardArrowDataAttributes["side"] = "data-side";
  /**
   * Present when the preview card arrow is uncentered.
   */
  PreviewCardArrowDataAttributes["uncetered"] = "data-uncentered";
  return PreviewCardArrowDataAttributes;
}({});
