// @base-ui-components/react/esm/menu/arrow/MenuArrowDataAttributes.js
import { CommonPopupDataAttributes } from '../../utils/popupStateMapping.js';
export let MenuArrowDataAttributes = function (MenuArrowDataAttributes) {
  /**
   * Present when the menu popup is open.
   */
  MenuArrowDataAttributes[MenuArrowDataAttributes["open"] = CommonPopupDataAttributes.open] = "open";
  /**
   * Present when the menu popup is closed.
   */
  MenuArrowDataAttributes[MenuArrowDataAttributes["closed"] = CommonPopupDataAttributes.closed] = "closed";
  /**
   * Present when the anchor is hidden.
   */
  MenuArrowDataAttributes[MenuArrowDataAttributes["anchorHidden"] = CommonPopupDataAttributes.anchorHidden] = "anchorHidden";
  /**
   * Indicates which side the menu is positioned relative to the trigger.
   * @type {'top' | 'bottom' | 'left' | 'right' | 'inline-end' | 'inline-start'}
   */
  MenuArrowDataAttributes["side"] = "data-side";
  /**
   * Present when the menu arrow is uncentered.
   */
  MenuArrowDataAttributes["uncetered"] = "data-uncentered";
  return MenuArrowDataAttributes;
}({});
