// @base-ui-components/react/esm/tooltip/popup/TooltipPopupDataAttributes.js
import { CommonPopupDataAttributes } from '../../utils/popupStateMapping.js';
export let TooltipPopupDataAttributes = function (TooltipPopupDataAttributes) {
  /**
   * Present when the tooltip is open.
   */
  TooltipPopupDataAttributes[TooltipPopupDataAttributes["open"] = CommonPopupDataAttributes.open] = "open";
  /**
   * Present when the tooltip is closed.
   */
  TooltipPopupDataAttributes[TooltipPopupDataAttributes["closed"] = CommonPopupDataAttributes.closed] = "closed";
  /**
   * Present when the tooltip is animating in.
   */
  TooltipPopupDataAttributes[TooltipPopupDataAttributes["startingStyle"] = CommonPopupDataAttributes.startingStyle] = "startingStyle";
  /**
   * Present when the tooltip is animating out.
   */
  TooltipPopupDataAttributes[TooltipPopupDataAttributes["endingStyle"] = CommonPopupDataAttributes.endingStyle] = "endingStyle";
  /**
   * Indicates which side the tooltip is positioned relative to the trigger.
   * @type {'top' | 'bottom' | 'left' | 'right' | 'inline-end' | 'inline-start'}
   */
  TooltipPopupDataAttributes["side"] = "data-side";
  /**
   * Present if animations should be instant.
   * @type {'delay' | 'dismiss' | 'focus'}
   */
  TooltipPopupDataAttributes["instant"] = "data-instant";
  return TooltipPopupDataAttributes;
}({});
